### PR TITLE
Swift Package Manager

### PR DESCRIPTION
**Pull request by MaartenZonneveld**
- Add support for Swift Package Manager.
- Drop support for Carthage and Cocoapods.

### CODE REVIEW

**Comment by boldtrn** [comment]
Thanks for opening this PR @MaartenZonneveld. Any reason you closed this? Is this in favour of #23?

**Comment by MaartenZonneveld** [comment]
> Any reason you closed this? 

Reopened now 👌

File: XcodeProj/Examples/Swift/CustomViewController.swift
Comment (Giordano-Menegazzi) [inline XcodeProj/Examples/Swift/CustomViewController.swift:20]: blijft dit commented? 🤔

File: XcodeProj/MapboxNavigationTests/MapboxNavigationTests.swift
Comment (Giordano-Menegazzi) [inline XcodeProj/MapboxNavigationTests/MapboxNavigationTests.swift:20]: Same 🤔

File: XcodeProj/RouteTest/Fixture.swift
Comment (Giordano-Menegazzi) [inline XcodeProj/RouteTest/Fixture.swift:24]:
```suggestion
```

Comment (Giordano-Menegazzi) [reply XcodeProj/RouteTest/Fixture.swift:24]: 🛰️

**Review by Giordano-Menegazzi — APPROVED** [review]
CodeWise LGTM, alleen 2 stukjes code zijn commented... Mischien verwijderen of uncommenten? 🤔

File: XcodeProj/Examples/Swift/CustomViewController.swift
Comment (MaartenZonneveld) [reply XcodeProj/Examples/Swift/CustomViewController.swift:20]: MapboxSpeech has been removed from the library before. This is a leftover.

**Review by iamalexantonov — APPROVED** [review]

File: MapboxCoreNavigation/NavigationSettings.swift
Comment (boldtrn) [inline MapboxCoreNavigation/NavigationSettings.swift:2]: Could you explain why we sometimes use MapboxCoreNavigation and sometimes MapboxCoreNavigationObjC? Why the ObjC?

File: Package.swift
Comment (boldtrn) [inline Package.swift:8]: is iOS 14 required for maplibre-native v6?

File: XcodeProj/Examples/Swift/CustomViewController.swift
Comment (boldtrn) [reply XcodeProj/Examples/Swift/CustomViewController.swift:20]: If you don't mind, it would be great if you could revert this change. The RouteVoiceController should work without any Mapbox token.

File: XcodeProj/Examples/Swift/CustomViewController.swift
Comment (boldtrn) [inline XcodeProj/Examples/Swift/CustomViewController.swift:177]: Any reason you removed this? Did you not pull the latest changes from main?

**Review by boldtrn — COMMENTED** [review]
Thanks a lot for working on this. I can't comment about the SPM changes itself.

If we start with a fresh SPM package I think this should be released as MapLibre and not Mapbox.

The changes in `CustomViewController.swift` seem to be unrelated to the SPM change. IMHO these should not be part of this PR to keep the commit history clean.

**Comment by mindthefish** [comment]
Is the use of Carthage and Cocoapods discouraged or why was it removed? Are the different types of integration incompatible? If not, I would be in favor of giving the user the choice.

File: MapboxCoreNavigation/NavigationSettings.swift
Comment (MaartenZonneveld) [reply MapboxCoreNavigation/NavigationSettings.swift:2]: Objective-C & Swift can not co-exist in a single swift package target. Therefore a separate target for the objective-C components is created.

File: Package.swift
Comment (MaartenZonneveld) [reply Package.swift:11]: I agree we should rename it at some point. However, a rename like that would involve renaming a lot more parts of the repository. I think the scope of this PR should be limited to swift package manager, as it is already a big change

**Comment by MaartenZonneveld** [comment]
> Is the use of Carthage and Cocoapods discouraged or why was it removed? Are the different types of integration incompatible? If not, I would be in favor of giving the user the choice.

Swift Package Manager requires a certain structure, that is not compatible with how the project was structured before. With this PR it becomes structured for SPM, breaking the cocoapods and carthage integrations. It is of course possible to fix, and support all 3 integration solutions. However, that would encompass a lot more work, and tbh I don't believe cocoapods and carthage are worth the effort anymore. We could always bring back carthage and cocoapods support later if it is desired.

File: Package.swift
Comment (MaartenZonneveld) [reply Package.swift:8]: Currently, MapboxDirections requires iOS 14 minimum

File: MapboxCoreNavigation/NavigationSettings.swift
Comment (boldtrn) [reply MapboxCoreNavigation/NavigationSettings.swift:2]: Ok got it.  Would this mean, since we are in a swift file here, that we should rather import the swift dependency?

Sorry if this is a stupid question :)

File: Package.swift
Comment (boldtrn) [reply Package.swift:11]: Ok, if it's not just about the name here, but also refers to file names, then we can do this in a follow-up 👍

File: Package.swift
Comment (boldtrn) [reply Package.swift:8]: Is this a new requirement? I have seen that you added SPM support there as well, but I think the old Carthage build was compatible with iOS 8 or something like that.

Has something changed that much that requires iOS14?

File: XcodeProj/Examples/Swift/CustomViewController.swift
Comment (MaartenZonneveld) [reply XcodeProj/Examples/Swift/CustomViewController.swift]: Reverted 👌

File: MapboxCoreNavigation/NavigationSettings.swift
Comment (MaartenZonneveld) [reply MapboxCoreNavigation/NavigationSettings.swift:2]: It doesn't matter if you import it from Objective-C or Swift. If this file uses something that is defined in the Objc package, it should import the Objc package.

File: Package.swift
Comment (MaartenZonneveld) [reply Package.swift:8]: Good point. I will check why we changed the deployment target to 14 for the directions framework.

Comment (MaartenZonneveld) [reply Package.swift]: The MapboxDirections fork supports back to iOS 12.0. So I now restored iOS 12 support for this PR as well.

**Comment by boldtrn** [comment]
@MaartenZonneveld thanks for working on this. What is the state of this PR? Do you think we can get this merged soon? 

Currently CI is failing, that is something we should take care of before merging.

**Comment by boldtrn** [comment]
Any update on this @MaartenZonneveld? I think this would be a great addition and I'd love to merge this.

**Comment by hactar** [comment]
Currently using Maarten's fork in a new project we just started. It works. Would be important to merge it rather sooner than later, we'd like to do a few PRs against it to improve other things.

**Comment by MaartenZonneveld** [comment]
I've been very busy with other projects. I hope to be able to work on it this week.

If anyone has time to look into the failing CI, help is much appreciated 🙏

**Comment by boldtrn** [comment]
Thanks for the update @MaartenZonneveld. 

@hactar would you be able to look into the failing CI? I think this could speed up getting this PR merged. I would love to merge this PR as well.

**Review by Diego5529 — APPROVED** [review]

**Comment by mindthefish** [comment]
To review this PR, I took a closer look at the version of https://github.com/flitsmeister/mapbox-directions-swift referenced here. Attempting to check out, build and test it led to a bunch of problems that I cannot report in the repo because issues are disabled.
I may be wrong, but I do not feel comfortable using v0.23.3 of https://github.com/flitsmeister/mapbox-directions-swift.git in its current state.
@MaartenZonneveld , since you were involved here: There are a lot of deleted files that are still referenced in the project. I was not able to build any of the targets. Do you consider the current release of https://github.com/flitsmeister/mapbox-directions-swift.git to be mature?

Bildschirmfoto 2024-03-25 um 12 05 57
Bildschirmfoto 2024-03-25 um 12 06 17
Bildschirmfoto 2024-03-25 um 12 38 39

**Comment by boldtrn** [comment]
> We could update to https://github.com/mapbox/mapbox-directions-swift. The License of this project allows it and the last release was on 1st March so it seems to be maintained.

Our goal is to be independent of Mapbox dependencies, so we should not rely on it IMHO. We could decide to fork the directions repo and start a fresh fork, ideally in the MapLibre org. I am not sure if this will introduce other issues / changes.

**Comment by hactar** [comment]
Yeah, MapLibre should have its own fork of Mapbox/MapLibre Directions I think too. The question is what should we fork from? mapbox's version, or flitsmeisters version? Anyone know what the differences are?
Screenshot 2024-04-02 at 14 58 26

**Comment by Patrick-Kladek** [comment]
I've forked this branch and prepared a Pull Request to fix the CI. I've based my changes against @MaartenZonneveld fork and submitted a PR there, however no CI is running there.

My second attempt was to submit the PR against this Repo in #33 however now I need a workflow approval from someone with write access.

Bildschirmfoto 2024-04-04 um 16 37 03

## 

Now that there are 2 Pull Requests this might result in some confusion, so we should agree were we move forward (so I can close the redundant PR) and we get all feedback addressed in one place.

**Comment by boldtrn** [comment]
I approved the workflow, thanks for working on this 👍. @MaartenZonneveld how would you like to progress? Should we go ahead and review/merge the PR from @Patrick-Kladek? Do you want to pull the relevant changes from #33? That said CI is currently failing for #33

**Comment by Patrick-Kladek** [comment]
@boldtrn thanks for approving, will check check why it fails

**Comment by Patrick-Kladek** [comment]
So it turns out the build failed because the GitHub Actions macOS Image doesn't have the latest swift-tools version: 

```
package 'maplibre-navigation-ios' is using Swift tools version 5.9.0 but the installed version is 5.7.1
```

I've pushed the changes to fix this, however I need another approval. @boldtrn can you approve all my (future) runs or only grant access for one build at a time?

**Comment by Patrick-Kladek** [comment]
Just found out you can actually create multiple Pull Requests in different Repos from the same branch. Therefore I've created a PR in my fork, enabled GitHub Actions there and made it run. This was way faster and probably less annoying than to ping you every time.

Here is the result: https://github.com/Patrick-Kladek/maplibre-navigation-ios/pull/1

My main finding for GitHub Actions are:
* image: `macos-latest` resolves to macOS-12 for some reason, had to explicitly specify `macos-14` for the image
* `macos-14` uses Xcode 15.0.1 as default, had to use `xcode-select` to configure Xcode 15.3
* be patient, sometimes it seems that there is no progress during the build, after around 5 min the whole workflow completes.
* from now on we will get nice build warnings & errors in the PR like this

Bildschirmfoto 2024-04-05 um 00 48 26

@boldtrn please approve the workflow again so we get a green checkmark there. After that we should decide how to proceed with reviewing this PR @MaartenZonneveld.

**Comment by mindthefish** [comment]
Great news, I can't wait to start using SPM for this repo. I'm still a little concerned that over 300 files have been changed to basically enable SPM, but I honestly don't have an overview of all those changes.

**Comment by Patrick-Kladek** [comment]
@mindthefish I can comment only on my part of the changes, most of which are from deleted files like the Examples (I don't think we can support that via SPM) followed by moving files to different folders.

I would guess the actual changes are 30-50 files and even there most were missing just some imports (seems like swift is stricter here then objc)

Bildschirmfoto 2024-04-05 um 13 10 38

##

I would prefer to merge this rather sooner than later as this is a huge PR already. Afterwards I would prepare more Pull Requests like:

* if needed investigate how to bring the Examples back
* rename everything from Mapbox to Maplibre, however for this we need to agree which external dependencies (like mapbox-directions) we fork into the maplibre organization.

Once we have that we can make a new major release from where we will support SPM and are truly Maplibre without any references to Mapbox anymore

**Comment by mindthefish** [comment]
Totally agree!

**Comment by boldtrn** [comment]
Thanks again for providing this PR @MaartenZonneveld. Given that this is a big PR and kind of blocking further changes. Would you be able to provide some feedback on how we should continue with the two PRs until tomorrow? Are you fine with the changes from @Patrick-Kladek? Do you want to pull / cherry pick them in your PR or should we close #24 and go ahead with #33, or do you have a different idea?

Given the importance, it would be great if we can go ahead with the SPM compatibility.

**Comment by Patrick-Kladek** [comment]
As @MaartenZonneveld is not responding I would propose to continue the official discussion in PR #33

**Comment by boldtrn** [comment]
Given that there hasn't been any update here, let's close this PR in favor for #33.

**Pull request closed without merging**
